### PR TITLE
Ir parser

### PR DESCRIPTION
**Pull request by livingshade**
Parser and part of AST for ADN IR
Refer to `./ir/readme.md` for usage

### CODE REVIEW

File: elements/ir/adaptive.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		}
+		False => {
+			send(err('ratelimit'), APP);
+		}
+	};
+}
+
+
+fn resp(rpc_resp) {
+    send_time := outstanding.get(resp.get(id));
```
Comment (Romero027) [inline elements/ir/adaptive.rs:27]: update_window is also not defined.

File: elements/ir/ratelimit.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+fn init() {
+	last := current_time();
+	limit := 50;
+	token := 5;
+	per_sec := 5.0;
+}
+
+fn req(rpc_req) {
+	token := min(lim, per_sec * (current_time() - last));
```
Comment (Romero027) [inline elements/ir/ratelimit.rs:16]: Fix indentation?

File: ir/frontend/ir.lark [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+quoted_string: /'[^']*'/
+
+err: "err" "(" quoted_string ")"
+
+%import common.SIGNED_NUMBER -> NUMBER
+%import common.WS
+%import common.CPP_COMMENT
+%import common.CNAME
+%ignore WS
+%ignore CPP_COMMENT
```
Comment (Romero027) [inline ir/frontend/ir.lark:78]: add built-in functions? (e.g., current_time() )

**Comment by livingshade** [comment]
@Romero027 I basically finished the AST and implemented a printer to print AST back to text.

**Review by Romero027 — APPROVED** [review]

**Pull request merged**